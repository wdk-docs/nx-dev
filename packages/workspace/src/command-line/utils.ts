import * as yargsParser from 'yargs-parser';
import * as yargs from 'yargs';
import { readNxJson } from '../core/file-utils';
import { output } from '../utilities/output';
import { names } from '@nrwl/devkit';
import type { NxAffectedConfig } from '@nrwl/devkit';

const runOne: string[] = [
  'target',
  'configuration',
  'prod',
  'runner',
  'parallel',
  'max-parallel',
  'exclude',
  'only-failed',
  'help',
  'with-deps',
  'skip-nx-cache',
  'scan',
  'hide-cached-output',
];

const runMany: string[] = [...runOne, 'projects', 'all'];

const runAffected: string[] = [
  ...runOne,
  'untracked',
  'uncommitted',
  'all',
  'base',
  'head',
  'files',
  'plain',
  'select',
];

export interface RawNxArgs extends NxArgs {
  prod?: boolean;
}

export interface NxArgs {
  target?: string;
  configuration?: string;
  runner?: string;
  parallel?: number;
  untracked?: boolean;
  uncommitted?: boolean;
  all?: boolean;
  base?: string;
  head?: string;
  exclude?: string[];
  files?: string[];
  onlyFailed?: boolean;
  'only-failed'?: boolean;
  verbose?: boolean;
  help?: boolean;
  version?: boolean;
  plain?: boolean;
  withDeps?: boolean;
  'with-deps'?: boolean;
  projects?: string[];
  select?: string;
  skipNxCache?: boolean;
  'skip-nx-cache'?: boolean;
  'hide-cached-output'?: boolean;
  hideCachedOutput?: boolean;
  scan?: boolean;
}

const ignoreArgs = ['$0', '_'];

export function splitArgsIntoNxArgsAndOverrides(
  args: yargs.Arguments,
  mode: 'run-one' | 'run-many' | 'affected' | 'print-affected',
  options = { printWarnings: true }
): { nxArgs: NxArgs; overrides: yargs.Arguments } {
  const nxSpecific =
    mode === 'run-one' ? runOne : mode === 'run-many' ? runMany : runAffected;

  const nxArgs: RawNxArgs = {};
  const overrides = yargsParser(args._ as string[], {
    configuration: {
      'strip-dashed': true,
      'dot-notation': false,
    },
  });
  // This removes the overrides from the nxArgs._
  args._ = overrides._;

  delete overrides._;

  Object.entries(args).forEach(([key, value]) => {
    const dasherized = names(key).fileName;
    if (nxSpecific.includes(dasherized) || dasherized.startsWith('nx-')) {
      if (value !== undefined) nxArgs[key] = value;
    } else if (!ignoreArgs.includes(key)) {
      overrides[key] = value;
    }
  });

  if (mode === 'run-many') {
    if (!nxArgs.projects) {
      nxArgs.projects = [];
    } else {
      nxArgs.projects = (args.projects as string)
        .split(',')
        .map((p: string) => p.trim());
    }
  }

  if (nxArgs.prod) {
    delete nxArgs.prod;
    nxArgs.configuration = 'production';
  }

  if (mode === 'affected') {
    if (options.printWarnings && nxArgs.all) {
      output.warn({
        title: `Running affected:* commands with --all can result in very slow builds.`,
        bodyLines: [
          `${output.bold(
            '--all'
          )} is not meant to be used for any sizable project or to be used in CI.`,
          '',
          `${output.colors.gray(
            'Learn more about checking only what is affected: '
          )}https://nx.dev/latest/angular/cli/affected#affected.`,
        ],
      });
    }

    if (
      !nxArgs.files &&
      !nxArgs.uncommitted &&
      !nxArgs.untracked &&
      !nxArgs.base &&
      !nxArgs.head &&
      !nxArgs.all &&
      args._.length >= 3
    ) {
      nxArgs.base = args._[1] as string;
      nxArgs.head = args._[2] as string;
    }

    // Allow setting base and head via environment variables (lower priority then direct command arguments)
    if (!nxArgs.base && process.env.NX_BASE) {
      nxArgs.base = process.env.NX_BASE;
      if (options.printWarnings) {
        output.note({
          title: `No explicit --base argument provided, but found environment variable NX_BASE so using its value as the affected base: ${output.bold(
            `${nxArgs.base}`
          )}`,
        });
      }
    }
    if (!nxArgs.head && process.env.NX_HEAD) {
      nxArgs.head = process.env.NX_HEAD;
      if (options.printWarnings) {
        output.note({
          title: `No explicit --head argument provided, but found environment variable NX_HEAD so using its value as the affected head: ${output.bold(
            `${nxArgs.head}`
          )}`,
        });
      }
    }

    if (!nxArgs.base) {
      const affectedConfig = getAffectedConfig();
      nxArgs.base = affectedConfig.defaultBase;

      // No user-provided arguments to set the affected criteria, so inform the user of the defaults being used
      if (
        options.printWarnings &&
        !nxArgs.head &&
        !nxArgs.files &&
        !nxArgs.uncommitted &&
        !nxArgs.untracked &&
        !nxArgs.all
      ) {
        output.note({
          title: `Affected criteria defaulted to --base=${output.bold(
            `${nxArgs.base}`
          )} --head=${output.bold('HEAD')}`,
        });
      }
    }
  }

  if (!nxArgs.skipNxCache) {
    nxArgs.skipNxCache = process.env.NX_SKIP_NX_CACHE === 'true';
  }

  if (args['parallel'] === 'false' || args['parallel'] === false) {
    nxArgs['parallel'] = 1;
  } else if (
    args['parallel'] === 'true' ||
    args['parallel'] === true ||
    args['parallel'] === ''
  ) {
    nxArgs['parallel'] = Number(
      nxArgs['maxParallel'] || nxArgs['max-parallel'] || 3
    );
  } else if (args['parallel'] !== undefined) {
    nxArgs['parallel'] = Number(args['parallel']);
  }

  return { nxArgs, overrides };
}

export function getAffectedConfig(): NxAffectedConfig {
  const config = readNxJson();

  return {
    defaultBase: config.affected?.defaultBase || 'main',
  };
}
