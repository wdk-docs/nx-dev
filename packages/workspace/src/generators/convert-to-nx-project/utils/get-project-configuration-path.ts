import { ProjectConfiguration } from '@nrwl/devkit';
import { join } from 'path';

export function getProjectConfigurationPath(
  configuration: ProjectConfiguration
) {
  return join(configuration.root, 'project.json');
}
