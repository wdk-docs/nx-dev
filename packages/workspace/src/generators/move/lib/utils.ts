import {
  getWorkspaceLayout,
  joinPathFragments,
  ProjectConfiguration,
  Tree,
} from '@nrwl/devkit';

import { Schema } from '../schema';

/**
 * This helper function ensures that we don't move libs or apps
 * outside of the folders they should be in.
 *
 * This will break if someone isn't using the default libs/apps
 * folders. In that case, they're on their own :/
 */
export function getDestination(
  host: Tree,
  schema: Schema,
  project: ProjectConfiguration
): string {
  const projectType = project.projectType;

  const workspaceLayout = getWorkspaceLayout(host);

  let rootFolder = workspaceLayout.libsDir;
  if (projectType === 'application') {
    rootFolder = workspaceLayout.appsDir;
  }
  return joinPathFragments(rootFolder, schema.destination);
}

/**
 * Replaces slashes with dashes
 *
 * @param path
 */
export function getNewProjectName(path: string): string {
  return path.replace(/\//g, '-');
}

/**
 * Normalizes slashes (removes duplicates)
 *
 * @param input
 */
export function normalizeSlashes(input: string): string {
  return input
    .split('/')
    .filter((x) => !!x)
    .join('/');
}
