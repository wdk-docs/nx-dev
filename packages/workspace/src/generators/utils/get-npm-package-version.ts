export function getNpmPackageVersion(packageName: string): string | null {
  try {
    const version = require('child_process').execSync(
      `npm view ${packageName} version`,
      { stdio: ['pipe', 'pipe', 'ignore'] }
    );

    if (version) {
      return version
        .toString()
        .trim()
        .replace(/^\n*|\n*$/g, '');
    }
  } catch (err) {}
  return null;
}
