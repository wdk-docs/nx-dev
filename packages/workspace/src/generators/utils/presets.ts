export enum Preset {
  Apps = 'apps',
  Empty = 'empty', // same as apps, deprecated
  Core = 'core',
  NPM = 'npm', // same as core, deprecated
  TS = 'ts',
  WebComponents = 'web-components',
  Angular = 'angular',
  AngularWithNest = 'angular-nest',
  React = 'react',
  ReactWithExpress = 'react-express',
  ReactNative = 'react-native',
  NextJs = 'next',
  Nest = 'nest',
  Express = 'express',
}
