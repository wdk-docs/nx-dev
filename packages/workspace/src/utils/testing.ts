import { join } from 'path';
import { SchematicTestRunner } from '@angular-devkit/schematics/testing';
import { Rule, Tree } from '@angular-devkit/schematics';
import { updateWorkspace } from './workspace';
import { TestingArchitectHost } from '@angular-devkit/architect/testing';
import { schema } from '@angular-devkit/core';
import { Architect } from '@angular-devkit/architect';
import { MockBuilderContext } from './testing-utils';
import { names } from '@nrwl/devkit';

const testRunner = new SchematicTestRunner(
  '@nrwl/workspace',
  join(__dirname, '../../generators.json')
);

testRunner.registerCollection(
  '@nrwl/jest',
  join(__dirname, '../../../jest/generators.json')
);

testRunner.registerCollection(
  '@nrwl/cypress',
  join(__dirname, '../../../cypress/generators.json')
);

testRunner.registerCollection(
  '@nrwl/express',
  join(__dirname, '../../../express/generators.json')
);

testRunner.registerCollection(
  '@nrwl/react',
  join(__dirname, '../../../react/generators.json')
);

testRunner.registerCollection(
  '@nrwl/angular',
  join(__dirname, '../../../angular/generators.json')
);

testRunner.registerCollection(
  '@nrwl/next',
  join(__dirname, '../../../next/generators.json')
);

testRunner.registerCollection(
  '@nrwl/node',
  join(__dirname, '../../../node/generators.json')
);

testRunner.registerCollection(
  '@nrwl/nest',
  join(__dirname, '../../../nest/generators.json')
);

testRunner.registerCollection(
  '@nrwl/web',
  join(__dirname, '../../../web/generators.json')
);

const migrationTestRunner = new SchematicTestRunner(
  '@nrwl/workspace/migrations',
  join(__dirname, '../../migrations.json')
);

export function runExternalSchematic<T = any>(
  collectionName: string,
  schematicName: string,
  options: T,
  tree: Tree
) {
  return testRunner
    .runExternalSchematicAsync(collectionName, schematicName, options, tree)
    .toPromise();
}

export function runSchematic<T = any>(
  schematicName: string,
  options: T,
  tree: Tree
) {
  return testRunner.runSchematicAsync(schematicName, options, tree).toPromise();
}

export function callRule(rule: Rule, tree: Tree) {
  return testRunner.callRule(rule, tree).toPromise();
}

export function runMigration(migrationName: string, options: any, tree: Tree) {
  return migrationTestRunner
    .runSchematicAsync(migrationName, options, tree)
    .toPromise();
}

export function createLibWithTests(
  tree: Tree,
  libName: string,
  testBuilder: string,
  testSetupFile: string
): Promise<Tree> {
  const { fileName } = names(libName);

  tree.create(`/libs/${fileName}/src/index.ts`, `\n`);

  return callRule(
    updateWorkspace((workspace) => {
      workspace.projects.add({
        name: fileName,
        root: `libs/${fileName}`,
        projectType: 'library',
        sourceRoot: `libs/${fileName}/src`,
        architect: {
          test: {
            builder: testBuilder,
            options: {
              setupFile: `libs/${fileName}/src/${testSetupFile}`,
            },
          },
        },
      });
    }),
    tree
  );
}

export async function getTestArchitect() {
  const architectHost = new TestingArchitectHost('/root', '/root');
  const registry = new schema.CoreSchemaRegistry();
  registry.addPostTransform(schema.transforms.addUndefinedDefaults);

  const architect = new Architect(architectHost, registry);

  await architectHost.addBuilderFromPackage(join(__dirname, '../..'));

  return [architect, architectHost] as [Architect, TestingArchitectHost];
}

export async function getMockContext() {
  const [architect, architectHost] = await getTestArchitect();

  const context = new MockBuilderContext(architect, architectHost);
  await context.addBuilderFromPackage(join(__dirname, '../..'));
  return context;
}
