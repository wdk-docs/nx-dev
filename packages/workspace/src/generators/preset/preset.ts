import {
  addDependenciesToPackageJson,
  convertNxGenerator,
  formatFiles,
  generateFiles,
  installPackagesTask,
  names,
  PackageManager,
  readWorkspaceConfiguration,
  Tree,
  updateJson,
  updateWorkspaceConfiguration,
} from '@nrwl/devkit';
import { Schema } from './schema';

import { libraryGenerator } from '../library/library';

import { insertImport } from '../utils/insert-import';
import { insertStatement } from '../utils/insert-statement';
import { Preset } from '../utils/presets';
import { join } from 'path';

export async function presetGenerator(tree: Tree, options: Schema) {
  options = normalizeOptions(options);
  await createPreset(tree, options);
  await formatFiles(tree);
  return () => {
    installPackagesTask(tree);
  };
}

export const presetSchematic = convertNxGenerator(presetGenerator);
export default presetGenerator;

async function createPreset(tree: Tree, options: Schema) {
  if (
    options.preset === Preset.Empty ||
    options.preset === Preset.Apps ||
    options.preset === Preset.TS
  ) {
    return;
  } else if (options.preset === Preset.Angular) {
    const {
      applicationGenerator: angularApplicationGenerator,
    } = require('@nrwl' + '/angular/generators');

    await angularApplicationGenerator(tree, {
      name: options.name,
      style: options.style,
      linter: options.linter,
      standaloneConfig: options.standaloneConfig,
    });
    setDefaultCollection(tree, '@nrwl/angular');
  } else if (options.preset === Preset.React) {
    const {
      applicationGenerator: reactApplicationGenerator,
    } = require('@nrwl' + '/react');

    await reactApplicationGenerator(tree, {
      name: options.name,
      style: options.style,
      linter: options.linter,
      standaloneConfig: options.standaloneConfig,
    });
    setDefaultCollection(tree, '@nrwl/react');
  } else if (options.preset === Preset.NextJs) {
    const { applicationGenerator: nextApplicationGenerator } = require('@nrwl' +
      '/next');

    await nextApplicationGenerator(tree, {
      name: options.name,
      style: options.style,
      linter: options.linter,
      standaloneConfig: options.standaloneConfig,
    });
    setDefaultCollection(tree, '@nrwl/next');
  } else if (options.preset === Preset.WebComponents) {
    const { applicationGenerator: webApplicationGenerator } = require('@nrwl' +
      '/web');

    await webApplicationGenerator(tree, {
      name: options.name,
      style: options.style,
      linter: options.linter,
      standaloneConfig: options.standaloneConfig,
    });
    addDependenciesToPackageJson(
      tree,
      {},
      {
        '@ungap/custom-elements': '0.1.6',
      }
    );
    addPolyfills(
      tree,
      `apps/${names(options.name).fileName}/src/polyfills.ts`,
      ['@ungap/custom-elements']
    );
    setDefaultCollection(tree, '@nrwl/web');
  } else if (options.preset === Preset.AngularWithNest) {
    const {
      applicationGenerator: angularApplicationGenerator,
    } = require('@nrwl' + '/angular/generators');
    const { applicationGenerator: nestApplicationGenerator } = require('@nrwl' +
      '/nest');

    await angularApplicationGenerator(tree, {
      name: options.name,
      style: options.style,
      linter: options.linter,
      skipFormat: true,
      standaloneConfig: options.standaloneConfig,
    });
    await nestApplicationGenerator(tree, {
      name: 'api',
      frontendProject: options.name,
      linter: options.linter,
      standaloneConfig: options.standaloneConfig,
    });
    await libraryGenerator(tree, {
      name: 'api-interfaces',
      unitTestRunner: 'none',
      linter: options.linter,
      standaloneConfig: options.standaloneConfig,
    });
    setDefaultCollection(tree, '@nrwl/angular');
    connectAngularAndNest(tree, options);
  } else if (options.preset === Preset.ReactWithExpress) {
    const {
      applicationGenerator: expressApplicationGenerator,
    } = require('@nrwl' + '/express');
    const {
      applicationGenerator: reactApplicationGenerator,
    } = require('@nrwl' + '/react');

    await reactApplicationGenerator(tree, {
      name: options.name,
      style: options.style,
      linter: options.linter,
      standaloneConfig: options.standaloneConfig,
    });
    await expressApplicationGenerator(tree, {
      name: 'api',
      frontendProject: options.name,
      linter: options.linter,
      standaloneConfig: options.standaloneConfig,
    });
    await libraryGenerator(tree, {
      name: 'api-interfaces',
      unitTestRunner: 'none',
      linter: options.linter,
      standaloneConfig: options.standaloneConfig,
    });
    setDefaultCollection(tree, '@nrwl/react');
    connectReactAndExpress(tree, options);
  } else if (options.preset === Preset.Nest) {
    const { applicationGenerator: nestApplicationGenerator } = require('@nrwl' +
      '/nest');

    await nestApplicationGenerator(tree, {
      name: options.name,
      linter: options.linter,
    });
    setDefaultCollection(tree, '@nrwl/nest');
  } else if (options.preset === Preset.Express) {
    const {
      applicationGenerator: expressApplicationGenerator,
    } = require('@nrwl' + '/express');
    await expressApplicationGenerator(tree, {
      name: options.name,
      linter: options.linter,
      standaloneConfig: options.standaloneConfig,
    });
    setDefaultCollection(tree, '@nrwl/express');
  } else if (options.preset === 'react-native') {
    const { reactNativeApplicationGenerator } = require('@nrwl' +
      '/react-native');
    await reactNativeApplicationGenerator(tree, {
      name: options.name,
      linter: options.linter,
      standaloneConfig: options.standaloneConfig,
      e2eTestRunner: 'detox',
    });
    setDefaultCollection(tree, '@nrwl/react-native');
  } else if (options.preset === Preset.Core || options.preset === Preset.NPM) {
    setupPackageManagerWorkspaces(tree, options);
  } else {
    throw new Error(`Invalid preset ${options.preset}`);
  }
}

function setupPackageManagerWorkspaces(tree: Tree, options: Schema) {
  if (options.packageManager === 'pnpm') {
    generateFiles(tree, join(__dirname, './files/pnpm-workspace'), '.', {});
  } else {
    updateJson(tree, 'package.json', (json) => {
      json.workspaces = ['packages/**'];
      return json;
    });
  }
}

function connectAngularAndNest(host: Tree, options: Schema) {
  const { insertNgModuleImport } = require('@nrwl' +
    '/angular/src/generators/utils');
  host.write(
    'libs/api-interfaces/src/lib/api-interfaces.ts',
    `export interface Message { message: string }`
  );

  const modulePath = `apps/${options.name}/src/app/app.module.ts`;

  insertImport(host, modulePath, 'HttpClientModule', '@angular/common/http');

  insertNgModuleImport(host, modulePath, 'HttpClientModule');

  const scope = options.npmScope;
  const style = options.style ?? 'css';
  host.write(
    `apps/${options.name}/src/app/app.component.ts`,
    `import { Component } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Message } from '@${scope}/api-interfaces';

@Component({
  selector: '${scope}-root',
  templateUrl: './app.component.html',
  styleUrls: ['./app.component.${style}']
})
export class AppComponent {
  hello$ = this.http.get<Message>('/api/hello')
  constructor(private http: HttpClient) {}
}
    `
  );

  host.write(
    `apps/${options.name}/src/app/app.component.spec.ts`,
    `import { Component } from '@angular/core';
import { TestBed, async } from '@angular/core/testing';
import { HttpClientModule } from '@angular/common/http';
import { AppComponent } from './app.component';

describe('AppComponent', () => {
  beforeEach(async(() => {
    TestBed.configureTestingModule({
      declarations: [AppComponent],
      imports: [HttpClientModule]
    }).compileComponents();
  }));

  it('should create the app', () => {
    const fixture = TestBed.createComponent(AppComponent);
    const app = fixture.debugElement.componentInstance;
    expect(app).toBeTruthy();
  });
});
    `
  );

  host.write(
    `apps/${options.name}/src/app/app.component.html`,
    `<div style="text-align:center">
  <h1>Welcome to ${options.name}!</h1>
  <img
    width="450"
    src="https://raw.githubusercontent.com/nrwl/nx/master/images/nx-logo.png"
    alt="Nx - Smart, Fast and Extensible Build System"
  />
</div>
<div>Message: {{ (hello$|async)|json }}</div>
    `
  );

  host.write(
    `apps/api/src/app/app.controller.ts`,
    `import { Controller, Get } from '@nestjs/common';

import { Message } from '@${scope}/api-interfaces';

import { AppService } from './app.service';

@Controller()
export class AppController {
  constructor(private readonly appService: AppService) {}

  @Get('hello')
  getData(): Message {
    return this.appService.getData();
  }
}
    `
  );

  host.write(
    `apps/api/src/app/app.service.ts`,
    `import { Injectable } from '@nestjs/common';
import { Message } from '@${scope}/api-interfaces';

@Injectable()
export class AppService {
  getData(): Message {
    return { message: 'Welcome to api!' };
  }
}
    `
  );
}

function connectReactAndExpress(host: Tree, options: Schema) {
  const scope = options.npmScope;
  host.write(
    'libs/api-interfaces/src/lib/api-interfaces.ts',
    `export interface Message { message: string }`
  );

  host.write(
    `apps/${options.name}/src/app/app.tsx`,
    `import React, { useEffect, useState } from 'react';
import { Message } from '@${scope}/api-interfaces';

export const App = () => {
  const [m, setMessage] = useState<Message>({ message: '' });

  useEffect(() => {
    fetch('/api')
      .then(r => r.json())
      .then(setMessage);
  }, []);

  return (
    <>
      <div style={{ textAlign: 'center' }}>
        <h1>Welcome to ${options.name}!</h1>
        <img
          width="450"
          src="https://raw.githubusercontent.com/nrwl/nx/master/images/nx-logo.png"
          alt="Nx - Smart, Fast and Extensible Build System"
        />
      </div>
      <div>{m.message}</div>
    </>
  );
};

export default App;
    `
  );

  host.write(
    `apps/${options.name}/src/app/app.spec.tsx`,
    `import { cleanup, getByText, render, waitFor } from '@testing-library/react';
import React from 'react';
import App from './app';

describe('App', () => {
  afterEach(() => {
    delete global['fetch'];
    cleanup();
  });

  it('should render successfully', async () => {
    global['fetch'] = jest.fn().mockResolvedValueOnce({
      json: () => ({
        message: 'my message'
      })
    });

    const { baseElement } = render(<App />);
    await waitFor(() => getByText(baseElement, 'my message'));
  });
});
    `
  );

  host.write(
    `apps/api/src/main.ts`,
    `import * as express from 'express';
import { Message } from '@${scope}/api-interfaces';

const app = express();

const greeting: Message = { message: 'Welcome to api!' };

app.get('/api', (req, res) => {
  res.send(greeting);
});

const port = process.env.port || 3333;
const server = app.listen(port, () => {
  console.log('Listening at http://localhost:' + port + '/api');
});
server.on('error', console.error);
    `
  );
}

function setDefaultCollection(tree: Tree, defaultCollection: string) {
  const workspaceConfiguration = readWorkspaceConfiguration(tree);
  updateWorkspaceConfiguration(tree, {
    ...workspaceConfiguration,
    cli: {
      ...(workspaceConfiguration.cli || {}),
      defaultCollection,
    },
  });
}

function addPolyfills(host: Tree, polyfillsPath: string, polyfills: string[]) {
  for (const polyfill of polyfills) {
    insertStatement(host, polyfillsPath, `import '${polyfill}';\n`);
  }
}

function normalizeOptions(options: Schema): Schema {
  options.name = names(options.name).fileName;
  return options;
}
