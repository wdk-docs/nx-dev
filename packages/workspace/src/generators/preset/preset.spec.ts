import { Tree, readJson, NxJsonConfiguration } from '@nrwl/devkit';
import { createTreeWithEmptyWorkspace } from '@nrwl/devkit/testing';
import { overrideCollectionResolutionForTesting } from '@nrwl/devkit/ngcli-adapter';
import { presetGenerator } from './preset';
import * as path from 'path';
import { Preset } from '../utils/presets';

describe('preset', () => {
  let tree: Tree;

  beforeEach(() => {
    tree = createTreeWithEmptyWorkspace();
    overrideCollectionResolutionForTesting({
      '@nrwl/workspace': path.join(
        __dirname,
        '../../../../workspace/generators.json'
      ),
      '@nrwl/angular': path.join(
        __dirname,
        '../../../../angular/generators.json'
      ),
      '@nrwl/linter': path.join(
        __dirname,
        '../../../../linter/generators.json'
      ),
      '@nrwl/nest': path.join(__dirname, '../../../../nest/generators.json'),
      '@nrwl/node': path.join(__dirname, '../../../../node/generators.json'),
      '@nrwl/jest': path.join(__dirname, '../../../../jest/generators.json'),
      '@nrwl/cypress': path.join(
        __dirname,
        '../../../../cypress/generators.json'
      ),
      '@nrwl/express': path.join(
        __dirname,
        '../../../../express/generators.json'
      ),
    });
  });

  afterEach(() => {
    overrideCollectionResolutionForTesting(null);
  });

  it(`should create files (preset = ${Preset.Angular})`, async () => {
    await presetGenerator(tree, {
      name: 'proj',
      preset: Preset.Angular,
      cli: 'nx',
      style: 'css',
      linter: 'eslint',
      standaloneConfig: false,
    });
    expect(tree.children('apps/proj')).toMatchSnapshot();
    expect(tree.children('apps/proj/src/')).toMatchSnapshot();
    expect(tree.children('apps/proj/src/app')).toMatchSnapshot();

    expect(
      readJson<NxJsonConfiguration>(tree, 'nx.json').cli.defaultCollection
    ).toBe('@nrwl/angular');
  });

  it(`should create files (preset = ${Preset.WebComponents})`, async () => {
    await presetGenerator(tree, {
      name: 'proj',
      preset: Preset.WebComponents,
      cli: 'nx',
      standaloneConfig: false,
    });
    expect(tree.exists('/apps/proj/src/main.ts')).toBe(true);
    expect(
      readJson<NxJsonConfiguration>(tree, 'nx.json').cli.defaultCollection
    ).toBe('@nrwl/web');
  });

  it(`should create files (preset = ${Preset.React})`, async () => {
    await presetGenerator(tree, {
      name: 'proj',
      preset: Preset.React,
      style: 'css',
      linter: 'eslint',
      cli: 'nx',
      standaloneConfig: false,
    });
    expect(tree.exists('/apps/proj/src/main.tsx')).toBe(true);
    expect(
      readJson<NxJsonConfiguration>(tree, 'nx.json').cli.defaultCollection
    ).toBe('@nrwl/react');
  });

  it(`should create files (preset = ${Preset.NextJs})`, async () => {
    await presetGenerator(tree, {
      name: 'proj',
      preset: Preset.NextJs,
      style: 'css',
      linter: 'eslint',
      cli: 'nx',
      standaloneConfig: false,
    });
    expect(tree.exists('/apps/proj/pages/index.tsx')).toBe(true);
    expect(
      readJson<NxJsonConfiguration>(tree, 'nx.json').cli.defaultCollection
    ).toBe('@nrwl/next');
  });

  it(`should create files (preset = ${Preset.AngularWithNest})`, async () => {
    await presetGenerator(tree, {
      name: 'proj',
      preset: Preset.AngularWithNest,
      style: 'css',
      linter: 'eslint',
      cli: 'nx',
      standaloneConfig: false,
    });

    expect(tree.exists('/apps/proj/src/app/app.component.ts')).toBe(true);
    expect(tree.exists('/apps/api/src/app/app.controller.ts')).toBe(true);
    expect(tree.exists('/libs/api-interfaces/src/lib/api-interfaces.ts')).toBe(
      true
    );
  });

  it(`should create files (preset = ${Preset.ReactWithExpress})`, async () => {
    await presetGenerator(tree, {
      name: 'proj',
      preset: Preset.ReactWithExpress,
      style: 'css',
      linter: 'eslint',
      cli: 'nx',
      standaloneConfig: false,
    });

    expect(tree.exists('/apps/proj/src/app/app.tsx')).toBe(true);
    expect(tree.exists('/libs/api-interfaces/src/lib/api-interfaces.ts')).toBe(
      true
    );
    expect(tree.exists('/apps/proj/.eslintrc.json')).toBe(true);
    expect(tree.exists('/apps/api/.eslintrc.json')).toBe(true);
    expect(tree.exists('/libs/api-interfaces/.eslintrc.json')).toBe(true);
  });

  it(`should create files (preset = ${Preset.Express})`, async () => {
    await presetGenerator(tree, {
      name: 'proj',
      preset: Preset.Express,
      linter: 'eslint',
      cli: 'nx',
      standaloneConfig: false,
    });

    expect(tree.exists('apps/proj/src/main.ts')).toBe(true);
    expect(tree.exists('apps/proj/.eslintrc.json')).toBe(true);
  });

  it('should create files (preset = react-native)', async () => {
    await presetGenerator(tree, {
      name: 'proj',
      preset: Preset.ReactNative,
      linter: 'eslint',
      cli: 'nx',
      standaloneConfig: false,
    });

    expect(tree.exists('/apps/proj/src/app/App.tsx')).toBe(true);
    expect(
      readJson<NxJsonConfiguration>(tree, 'nx.json').cli.defaultCollection
    ).toBe('@nrwl/react-native');
  });

  describe('core preset', () => {
    describe('package manager workspaces', () => {
      it('should be configured in package.json', async () => {
        await presetGenerator(tree, {
          name: 'proj',
          preset: Preset.Core,
          linter: 'eslint',
          cli: 'nx',
          standaloneConfig: false,
          packageManager: 'npm',
        });

        expect(readJson(tree, 'package.json').workspaces)
          .toMatchInlineSnapshot(`
          Array [
            "packages/**",
          ]
        `);
      });

      it('should be configured in pnpm-workspace.yaml', async () => {
        await presetGenerator(tree, {
          name: 'proj',
          preset: Preset.Core,
          linter: 'eslint',
          cli: 'nx',
          standaloneConfig: false,
          packageManager: 'pnpm',
        });

        expect(tree.read('pnpm-workspace.yaml', 'utf-8'))
          .toMatchInlineSnapshot(`
          "packages:
            - 'packages/**'
          "
        `);
      });
    });
  });
});
