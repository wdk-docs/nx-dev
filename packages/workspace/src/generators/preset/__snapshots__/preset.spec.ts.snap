// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`preset should create files (preset = angular) 1`] = `
Array [
  "src",
  "tsconfig.editor.json",
  "tsconfig.json",
  ".browserslistrc",
  "tsconfig.app.json",
  "tsconfig.spec.json",
  ".eslintrc.json",
  "jest.config.js",
]
`;

exports[`preset should create files (preset = angular) 2`] = `
Array [
  "favicon.ico",
  "index.html",
  "main.ts",
  "polyfills.ts",
  "styles.css",
  "assets",
  "environments",
  "app",
  "test-setup.ts",
]
`;

exports[`preset should create files (preset = angular) 3`] = `
Array [
  "app.module.ts",
  "app.component.css",
  "app.component.html",
  "app.component.spec.ts",
  "app.component.ts",
  "nx-welcome.component.ts",
]
`;
